core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1070,6 +1071,70 @@ private ScanToCalcUnifyRule() {
     }
   }
 
+  /**
+   * A {@link SubstitutionVisitor.UnifyRule} that matches a {@link MutableCalc}
+   * which has {@link MutableAggregate} as  child.
+   * We try to compensate the {@link MutableCalc},
+   * then match the query to the top of {@link MutableCalc} in target.
+   */
+  private static class NothingToCalcUnifyRule extends AbstractUnifyRule {
+
+    public static final NothingToCalcUnifyRule INSTANCE =
+        new NothingToCalcUnifyRule();
+
+    private NothingToCalcUnifyRule() {
+      super(any(MutableRel.class),
+          any(MutableCalc.class), 0);
+    }
+
+    public UnifyResult apply(UnifyRuleCall call) {
+      final MutableRel rel = call.query;
+      final MutableCalc target = (MutableCalc) call.target;
+      final MutableRel targetInput = target.getInput();
+      final Pair<RexNode, List<RexNode>> qInputExplained = explainCalc(target);
+      if (rel.getParent() != null || target.getParent() != null) {
+        return null;
+      }
+      final Mappings.TargetMapping mapping =
+          Project.getMapping(target.getInput().rowType.getFieldCount(), qInputExplained.right);
+      if (mapping == null || Mappings.keepsOrdering(mapping)) {
+        return null;
+      }
+      MutableRel equalRel = rel.getInputs().get(0);
+      while (equalRel != targetInput) {
```

Comment:
I think the equality checking of input should rely on the frame work, not check by yourself.
If the matching pattern is as below 
```
 super(operand(query(0)),
          operand(MutableCalc.class, target(0)), 1);
```
SubstutitionVisitor will check the if `query` equals to the input of `target` for you